java/tools/src/main/java/org/apache/arrow/tools/EchoServer.java

```diff
@@ -135,6 +135,7 @@ public void run() throws IOException {
         Preconditions.checkState(reader.bytesRead() == writer.bytesWritten());
         LOGGER.debug(String.format("Echoed %d records", echoed));
         reader.close(false);
+        writer.close();
```

Comment:
The `ArrowStreamWriter` not being closed meant that the copies of dictionary vectors it holds weren't being closed, so the allocator was throwing an `IllegalStateException` due to outstanding buffers. This is a behaviour change that users might need to watch out for.

Reply:
Thanks. I think it's reasonable to expect that you must close the writer anyways (and that you should generally try-with-resources these writers).